### ISSUE TITLE
'error CS0165: Use of unassigned local variable' is not produced for structs with a private reference type field from a different assembly

### ISSUE THREAD

**Issue by hmemcpy** [dotnet/roslyn#1722]
Here's a minimal repro:

ClassLibrary1: **MyStruct.cs**
```cs
using System;

namespace ClassLibrary1
{
  public struct MyStruct
  {
    private readonly object a;
  }
}
```

ConsoleApplication1: **Program.cs**
```cs
using System;
using ClassLibrary1;

namespace ConsoleApplication1
{
  class Program
  {
    static void Main(string[] args)
    {
      MyStruct mystruct;

      Console.WriteLine(mystruct); //  <--MISSING: error CS0165 does not happen, project compiles in VS2015 CTP6.
    }
  }
}
```

In both VS2012 and VS2013 this produces an error, as expected. In VS2015 CTP6 this compiles successfully.
A struct with a *private* reference type field will not cause the compilation to fail. Changing to a value type, or making the field public will cause the error, as expected. Also, if the struct is defined in the same assembly, it will fail, as expected.

**Comment by hmemcpy** [comment]
Just tried it with `/features:strict`, and it gives an error, as with previous compilers.

**Comment by hmemcpy** [comment]
I just looked at some tests. It seems that here is a test that checks for this case exactly, but it is either incomplete, or makes sure there are NO errors produced:
https://github.com/dotnet/roslyn/blame/5fc7af63afb92566381ad44566f987cd6fedfe21/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/StructTests.cs#L364-L387

@gafter, "Blame" puts your name there :)

**Comment by gafter** [comment]
@hmemcpy This was a very painful but intentional decision. This duplicates the (buggy) behavior of the previous compiler. I strongly recommend you add the compiler flag `/feature:strict` to get the correct, specification-required (but not backward-compatible) behavior. This is not currently in the docs at https://github.com/dotnet/roslyn/tree/master/docs/compilers and it should be... I won't close this issue (as by design) until I've added docs explaining it.

Most reference assemblies omit private fields, so that you won't get these errors when compiling against a reference assembly. For that I blame whoever made your reference assembly. We plan to make the Roslyn compilers capable of producing reference assemblies that do not have this bug.

**Comment by hmemcpy** [comment]
@gafter thanks very much for the explanation!
It all started after ReSharper highlighted (which, I assumed, correctly) that a variable `Guid guid` was used without initialization, but it compiled fine.
Our product is a C#-to-JavaScript compiler (DuoCode), we build our own compiler and "mscorlib", so the Guid reference was from our reference assembly.
It didn't repro on a plain old C# project, until I realized this was only with private reference fields.

I guess this is now a ReSharper bug :)

**Comment by VSadov** [comment]
Old CSC had inconsistent behavior here. In some cases it would not load private fields as an optimization and that would result in more lenient definite assignment verification. Implementing the spec precisely in Roslyn was a breaking change and in some cases an adoption blocker. Therefore in the default (not strict) mode Roslyn compiler tries to imitate the behavior of the old compiler. It may not match that behavior precisely in all cases but should not be more restrictive to not break previously compiling code.

**Comment by gafter** [comment]
@VSadov Please "Verify"
